feat: use yarn-berry-deduplicate in yarn post-update

christophehurpeau opened the pull request:
<!-- If this is your first pull request: sign the CLA with this GitHub app: https://cla-assistant.io/renovatebot/renovate -->
<!-- Make sure the `Allow edits and access to secrets by maintainers` checkbox is checked on this pull request. -->
<!-- Please read https://github.com/renovatebot/renovate/blob/main/.github/contributing.md before you create your pull request.-->

## Changes

yarn-berry-deduplicate is a fork of yarn-deduplicate that works with the `fewer` strategy https://www.npmjs.com/package/yarn-berry-deduplicate

## Context

deduping with fewer strategy worked with yarn 1 but currently no longer works with yarn berry (issue: https://github.com/yarnpkg/berry/issues/2297)

## Documentation (please check one with an [x])

- [ ] I have updated the documentation, or
- [x] No documentation update is required

## How I've tested my work (please select one)

I have verified these changes via:

- [ ] Code inspection only, or
- [ ] Newly added/modified unit tests, or
- [x] No unit tests but ran on a real repository, or
- [ ] Both unit tests + ran on a real repository

<!-- Do you have any suggestions about this PR template? Edit it here: https://github.com/renovatebot/renovate/edit/main/.github/pull_request_template.md -->

<!-- Please do not force push to your PR's branch after you have created your PR, as doing so forces us to review the whole PR again. This makes it harder for us to review your work because we don't know what has changed. -->
<!-- PRs will always be squashed by us when we merge your work. Commit as many times as you need in this branch. -->

rarkins commented on lib/modules/manager/npm/post-update/yarn.ts:
```diff
@@ -248,7 +248,9 @@ export async function generateLockFile(
           commands.push(`yarn dedupe --strategy ${s}${cmdOptions}`);
         } else {
-          logger.debug(`yarn dedupe ${s} not available`);
+          commands.push(`npx yarn-berry-deduplicate --strategy ${s}`);
```
Why not the `yarn dlx` approach?

christophehurpeau replied:
yes ! I just copied/pasted without thinking the line above
```suggestion
          commands.push(`yarn dlx yarn-berry-deduplicate --strategy ${s}`);
```

viceice commented on lib/modules/manager/npm/post-update/yarn.ts:
```diff
@@ -248,7 +248,9 @@ export async function generateLockFile(
           commands.push(`yarn dedupe --strategy ${s}${cmdOptions}`);
         } else {
-          logger.debug(`yarn dedupe ${s} not available`);
+          commands.push(`yarn dlx yarn-berry-deduplicate --strategy ${s}`);
```
no dynamic download, create a tool constraint instead

rarkins replied:
FYI https://github.com/christophehurpeau/renovate/blob/17ba90f2fc3f61aade9f99a9be6300fb8692312f/lib/modules/manager/npm/post-update/yarn.ts#L244-L245

christophehurpeau replied:
How should I do that ?

viceice replied:
like the yarn command

viceice replied:
add the tool here like yarn

https://github.com/renovatebot/renovate/blob/de3cb95e91569768608df6a10d7e166f221c1243/lib/util/exec/containerbase.ts#L173-L177

add the tool to this array when needed

https://github.com/renovatebot/renovate/blob/de3cb95e91569768608df6a10d7e166f221c1243/lib/modules/manager/npm/post-update/yarn.ts#L101

christophehurpeau replied:
I'm not sure I understand how that would work. Would this is supposed to work with `yarn-deduplicate` also ? If so, maybe you could show how that would be refactored with that existing tool so I could make the same for this PR ?

viceice replied:
sure, but probably only next week

rarkins replied:
@viceice is this still todo?

viceice reviewed (approved):
because of time we can merge like now and refactor later. needs new test?

rarkins commented:
@christophehurpeau do you have time to add tests?

viceice reviewed (commented):
should have a test

christophehurpeau commented:
@viceice @rarkins I've updated the snapshots and added additional tests.
I've also improved calls to `yarn dlx yarn-berry-deduplicate` and `yarn install` to remove those that are unecessary. Passing both strategies will call `yarn dlx yarn-berry-deduplicate` without options as both the strategies will be applied immediatly.

viceice commented on lib/modules/manager/npm/post-update/yarn.ts:
```diff
@@ -252,10 +252,40 @@ export async function generateLockFile(
+            }`
+          );
+
+          if (!hasBothStrategies || !isYarnDedupeAvailable) {
```
are you sure about this condition? i think that needs to be removed and yarn install should always be called after dedupe.

christophehurpeau replied:
However I can remove if you're more comfortable without, it will mean more calls to yarn install even if not necessary

rarkins commented:
Closing due to inactivity
